packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart

```diff
@@ -0,0 +1,511 @@
+                  return _InspectorTreeRowWidget(
+                    key: PageStorageKey(row?.node),
+                    inspectorTreeState: this,
+                    row: row,
+                  );
+                },
+                childCount: controller.numRows,
+              ),
+              controller: _scrollControllerY,
+            ),
+          ),
+        ),
+      ),
+    );
+  }
+}
+
+final _defaultPaint = Paint()
+  ..color = defaultTreeLineColor
+  ..strokeWidth = chartLineStrokeWidth;
+
+/// Custom painter that draws lines indicating how parent and child rows are
+/// connected to each other.
+///
+/// Each rows object contains a list of ticks that indicate the x coordinates of
+/// vertical lines connecting other rows need to be drawn within the vertical
+/// area of the current row. This approach has the advantage that a row contains
+/// all information required to render all content within it but has the
+/// disadvantage that the x coordinates of each line connecting rows must be
+/// computed in advance.
+class _RowPainter extends CustomPainter {
+  _RowPainter(this.row, this._controller);
+
+  final InspectorTreeController _controller;
+  final InspectorTreeRow row;
+
+  @override
+  void paint(Canvas canvas, Size size) {
+    double currentX = 0;
+
+    if (row == null) {
+      return;
+    }
+    final InspectorTreeNode node = row.node;
+    final bool showExpandCollapse = node.showExpandCollapse;
+    for (int tick in row.ticks) {
+      currentX = _controller.getDepthIndent(tick) - columnWidth * 0.5;
+      // Draw a vertical line for each tick identifying a connection between
+      // an ancestor of this node and some other node in the tree.
+      canvas.drawLine(
+        Offset(currentX, 0.0),
+        Offset(currentX, rowHeight),
+        _defaultPaint,
+      );
+    }
+    // If this row is itself connected to a parent then draw the L shaped line
+    // to make that connection.
+    if (row.lineToParent) {
+      final paint = _defaultPaint;
+      currentX = _controller.getDepthIndent(row.depth - 1) - columnWidth * 0.5;
+      final double width = showExpandCollapse ? columnWidth * 0.5 : columnWidth;
+      canvas.drawLine(
+        Offset(currentX, 0.0),
+        Offset(currentX, rowHeight * 0.5),
+        paint,
+      );
+      canvas.drawLine(
+        Offset(currentX, rowHeight * 0.5),
+        Offset(currentX + width, rowHeight * 0.5),
+        paint,
+      );
+    }
+  }
+
+  @override
+  bool shouldRepaint(CustomPainter oldDelegate) {
+    // TODO(jacobr): check whether the row has different ticks.
+    return false;
+  }
+}
+
+/// Widget defining the contents of a single row in the InspectorTree.
+///
+/// This class defines the scaffolding around the rendering of the actual
+/// content of a [RemoteDiagnosticsNode] provided by
+/// [DiagnosticsNodeDescription] to provide a tree implementation with lines
+/// drawn between parent and child nodes when nodes have multiple children.
```

Comment:
nit: this documentation is a bit hard to parse. Breaking this up into multiple sentences would help.